packages/adblocker/src/engine/engine.ts

```diff
@@ -86,16 +86,72 @@ export interface Caching {
   write: (path: string, buffer: Uint8Array) => Promise<void>;
 }
 
-export default class FilterEngine extends EventEmitter<
-  | 'csp-injected'
-  | 'html-filtered'
-  | 'request-allowed'
-  | 'request-blocked'
-  | 'request-redirected'
-  | 'request-whitelisted'
-  | 'script-injected'
-  | 'style-injected'
-> {
+// Cosmetic rule is commonly used word in the project,
+// but cosmetic filter is used for better documentation.
+export type NetworkFilterMatchingContext = {
+  request: Request;
+
+  // `type` will cause a conflict in the variable name of `Request`.
+  matchType: FilterType.NETWORK;
+};
+
+// We do have a full context in case of the network filter matching.
+// It's because network filter matching does rely on an argument of the request.
+type CosmeticFilterMatchingContextBase = {
+  url: string;
+  domain: string | null | undefined;
+  hostname: string | undefined;
+
+  // Additional context given from user
+  matchType: FilterType.COSMETIC;
+  reference: any;
+};
+
+export type CosmeticFilterMatchingContext = CosmeticFilterMatchingContextBase &
+  Partial<{
+    classes: string[] | undefined;
+    hrefs: string[] | undefined;
+    ids: string[] | undefined;
+
+    allowGenericHides: boolean;
+    allowSpecificHides: boolean;
+
+    getBaseRules: boolean;
+    getInjectionRules: boolean;
+    getExtendedRules: boolean;
+    getRulesFromDOM: boolean;
+    getRulesFromHostname: boolean;
+  }>;
+
+export type MatchingContext = CosmeticFilterMatchingContext | NetworkFilterMatchingContext;
+
+type NetworkFilterMatchEvent = (request: Request, result: BlockingResponse) => void;
+type CosmeticInjectionEvent = (script: string, url: string) => void;
+
+export type EngineEventHandlers = {
+  'request-allowed': NetworkFilterMatchEvent;
+  'request-blocked': NetworkFilterMatchEvent;
+  'request-redirected': NetworkFilterMatchEvent;
+  'request-whitelisted': NetworkFilterMatchEvent;
+  'html-filtered': (
+    htmlSelectors: HTMLSelector[],
+    url: string,
+    context: CosmeticFilterMatchingContext,
+  ) => void;
+  'csp-injected': (csps: string, request: Request) => void;
+  'script-injected': CosmeticInjectionEvent;
+  'style-injected': CosmeticInjectionEvent;
+
+  // The below event kinds describe the internal process that was blackboxed before.
+  // In favor of backward-compatibility and the notification of the final action, above event kinds will be kept.
+  // We would match all *supported* filter by the engine.
+  // The result of *matched* filter can be differed by the end-user capability.
+  // However, we should rely on the given capability information and try to be accurate on our end as well.
+  // The unsupported filter won't be fired via the following events.
```

Comment:
Nit. I am a bit confused by this comment. Can you help me understand what point we are trying to get across here with this comment? I think some of this makes sense in the context of this PR but we need to keep in mind that this comment needs to make sense outside of this PR context so I would stick to describing what `filter-matched` does, and not refer to before/after.

Reply:
The comment was updated to only describe about its purpose and behavior.